Flink lookup will Intermittent timeout when Fluss cluster upgrading

### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss/issues) and found nothing similar.


### Fluss version

0.8.0 (latest release)

### Please describe the bug 🐞

Flink lookup will Intermittent timeout when Fluss cluster upgrading. Once a timeout occurs, it causes the Flink job to fail. This cannot be avoided no matter how large the `table.exec.async-lookup.timeout` is set.

The error is as follow:
```
java.lang.Exception: Could not complete the stream element: Record @ (undef) : +I(xxx)
	at org.apache.flink.streaming.api.operators.async.AsyncWaitOperator$ResultHandler.completeExceptionally(AsyncWaitOperator.java:636)
	at org.apache.flink.streaming.api.functions.async.AsyncFunction.timeout(AsyncFunction.java:97)
	at org.apache.flink.streaming.api.operators.async.AsyncWaitOperator$ResultHandler.timerTriggered(AsyncWaitOperator.java:654)
	at org.apache.flink.streaming.api.operators.async.AsyncWaitOperator$ResultHandler.lambda$registerTimeout$1(AsyncWaitOperator.java:649)
	at org.apache.flink.streaming.api.operators.async.AsyncWaitOperator.lambda$registerTimer$2(AsyncWaitOperator.java:433)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invokeProcessingTimeCallback(StreamTask.java:2186)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$deferCallbackToMailbox$27(StreamTask.java:2177)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$1.runThrowing(StreamTaskActionExecutor.java:50)
	at org.apache.flink.streaming.runtime.tasks.mailbox.Mail.run(Mail.java:101)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMail(MailboxProcessor.java:414)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.processMailsWhenDefaultActionUnavailable(MailboxProcessor.java:383)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.processMail(MailboxProcessor.java:368)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:229)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:1202)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:1146)
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:976)
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:955)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:768)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:580)
	at java.base/java.lang.Thread.run(Thread.java:991)
Caused by: java.util.concurrent.TimeoutException: Async function call has timed out.
	... 19 more
```

The root cause is still unknown, but there are two likely possibilities:
1. During upgrades, pods are recreated and their IP addresses change, which may cause metadata requests to take longer.
2. The Netty connection timeout is set to 120 seconds (`client.connect-timeout`). If the client sends a request to an IP that no longer exists—but previously had an established connection—it may wait for the full 120 seconds before timing out.

### Solution

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/fluss-client/src/main/java/org/apache/fluss/client/lookup/LookupQueue.java
+++ b/fluss-client/src/main/java/org/apache/fluss/client/lookup/LookupQueue.java
@@ -18,6 +18,7 @@
 package org.apache.fluss.client.lookup;
 
 import org.apache.fluss.annotation.Internal;
+import org.apache.fluss.annotation.VisibleForTesting;
 import org.apache.fluss.config.ConfigOptions;
 import org.apache.fluss.config.Configuration;
 
@@ -26,6 +27,8 @@ import javax.annotation.concurrent.ThreadSafe;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
 
 /**
@@ -38,13 +41,18 @@ class LookupQueue {
 
     private volatile boolean closed;
     // buffering both the Lookup and PrefixLookup.
+    // TODO This queue could be refactored into a memory-managed queue similar to
+    // RecordAccumulator, which would significantly improve the efficiency of lookup batching. Trace
+    // by https://github.com/apache/fluss/issues/2124
     private final ArrayBlockingQueue<AbstractLookupQuery<?>> lookupQueue;
+    private final BlockingQueue<AbstractLookupQuery<?>> reEnqueuedLookupQueue;
     private final int maxBatchSize;
     private final long batchTimeoutNanos;
 
     LookupQueue(Configuration conf) {
         this.lookupQueue =
                 new ArrayBlockingQueue<>(conf.get(ConfigOptions.CLIENT_LOOKUP_QUEUE_SIZE));
+        this.reEnqueuedLookupQueue = new LinkedBlockingQueue<>();
         this.maxBatchSize = conf.get(ConfigOptions.CLIENT_LOOKUP_MAX_BATCH_SIZE);
         this.batchTimeoutNanos = conf.get(ConfigOptions.CLIENT_LOOKUP_BATCH_TIMEOUT).toNanos();
         this.closed = false;
@@ -63,8 +71,21 @@ class LookupQueue {
         }
     }
 
+    void reEnqueue(AbstractLookupQuery<?> lookup) {
+        if (closed) {
+            throw new IllegalStateException(
+                    "Can not re-enqueue lookup operation since the LookupQueue is closed.");
+        }
+
+        try {
+            reEnqueuedLookupQueue.put(lookup);
+        } catch (InterruptedException e) {
+            lookup.future().completeExceptionally(e);
+        }
+    }
+
     boolean hasUnDrained() {
-        return !lookupQueue.isEmpty();
+        return !lookupQueue.isEmpty() || !reEnqueuedLookupQueue.isEmpty();
     }
 
     /** Drain a batch of {@link LookupQuery}s from the lookup queue. */
@@ -76,6 +97,16 @@ class LookupQueue {
             long waitNanos = batchTimeoutNanos - (System.nanoTime() - startNanos);
             if (waitNanos <= 0) {
                 break;
+            }
+
+            while (!reEnqueuedLookupQueue.isEmpty() && count < maxBatchSize) {
+                AbstractLookupQuery<?> lookup =
+                        reEnqueuedLookupQueue.poll(waitNanos, TimeUnit.NANOSECONDS);
+                if (lookup == null) {
+                    break;
+                }
+                lookupOperations.add(lookup);
+                count++;
             }
 
             AbstractLookupQuery<?> lookup = lookupQueue.poll(waitNanos, TimeUnit.NANOSECONDS);
@@ -97,10 +128,21 @@ class LookupQueue {
     List<AbstractLookupQuery<?>> drainAll() {
         List<AbstractLookupQuery<?>> lookupOperations = new ArrayList<>(lookupQueue.size());
         lookupQueue.drainTo(lookupOperations);
+        reEnqueuedLookupQueue.drainTo(lookupOperations);
         return lookupOperations;
     }
 
     public void close() {
         closed = true;
+    }
+
+    @VisibleForTesting
+    ArrayBlockingQueue<AbstractLookupQuery<?>> getLookupQueue() {
+        return lookupQueue;
+    }
+
+    @VisibleForTesting
+    BlockingQueue<AbstractLookupQuery<?>> getReEnqueuedLookupQueue() {
+        return reEnqueuedLookupQueue;
     }
 }

--- a/fluss-client/src/main/java/org/apache/fluss/client/lookup/LookupSender.java
+++ b/fluss-client/src/main/java/org/apache/fluss/client/lookup/LookupSender.java
@@ -157,6 +157,9 @@ class LookupSender implements Runnable {
             // lookup the leader node
             TableBucket tb = lookup.tableBucket();
             try {
+                // TODO Metadata requests are being sent too frequently here. consider first
+                // collecting the tables that need to be updated and then sending them together in
+                // one request.
                 leader = metadataUpdater.leaderFor(lookup.tablePath(), tb);
             } catch (Exception e) {
                 // if leader is not found, re-enqueue the lookup to send again.
@@ -423,7 +426,7 @@ class LookupSender implements Runnable {
     }
 
     private void reEnqueueLookup(AbstractLookupQuery<?> lookup) {
-        lookupQueue.appendLookup(lookup);
+        lookupQueue.reEnqueue(lookup);
     }
 
     private boolean canRetry(AbstractLookupQuery<?> lookup, Exception exception) {

--- a/fluss-client/src/test/java/org/apache/fluss/client/lookup/LookupQueueTest.java
+++ b/fluss-client/src/test/java/org/apache/fluss/client/lookup/LookupQueueTest.java
@@ -22,8 +22,12 @@ import org.apache.fluss.metadata.TableBucket;
 
 import org.junit.jupiter.api.Test;
 
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
 import static org.apache.fluss.config.ConfigOptions.CLIENT_LOOKUP_BATCH_TIMEOUT;
 import static org.apache.fluss.config.ConfigOptions.CLIENT_LOOKUP_MAX_BATCH_SIZE;
+import static org.apache.fluss.config.ConfigOptions.CLIENT_LOOKUP_QUEUE_SIZE;
 import static org.apache.fluss.record.TestData.DATA1_TABLE_PATH_PK;
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -56,6 +60,61 @@ class LookupQueueTest {
         assertThat(queue.drain()).hasSize(10);
         assertThat(queue.hasUnDrained()).isTrue();
         assertThat(queue.drainAll()).hasSize(10);
+        assertThat(queue.hasUnDrained()).isFalse();
+    }
+
+    @Test
+    void testAppendLookupBlocksWhenQueueIsFull() throws Exception {
+        Configuration conf = new Configuration();
+        conf.set(CLIENT_LOOKUP_QUEUE_SIZE, 5);
+        LookupQueue queue = new LookupQueue(conf);
+
+        appendLookups(queue, 5);
+        assertThat(queue.getLookupQueue()).hasSize(5);
+
+        CompletableFuture<Void> future =
+                CompletableFuture.runAsync(
+                        () -> {
+                            appendLookups(queue, 1); // will be blocked.
+                        });
+
+        // appendLookup should block and not complete immediately.
+        assertThat(future.isDone()).isFalse();
+
+        Thread.sleep(100);
+        // Still blocked after 100ms.
+        assertThat(future.isDone()).isFalse();
+
+        queue.drain();
+        future.get(1, TimeUnit.SECONDS);
+        assertThat(future.isDone()).isTrue();
+    }
+
+    @Test
+    void testReEnqueueNotBlock() throws Exception {
+        Configuration conf = new Configuration();
+        conf.set(CLIENT_LOOKUP_QUEUE_SIZE, 5);
+        conf.set(CLIENT_LOOKUP_MAX_BATCH_SIZE, 5);
+        LookupQueue queue = new LookupQueue(conf);
+
+        appendLookups(queue, 5);
+        assertThat(queue.getLookupQueue()).hasSize(5);
+        assertThat(queue.getReEnqueuedLookupQueue()).hasSize(0);
+
+        queue.reEnqueue(
+                new LookupQuery(DATA1_TABLE_PATH_PK, new TableBucket(1, 1), new byte[] {0}));
+        assertThat(queue.getLookupQueue()).hasSize(5);
+        // This batch will be put into re-enqueued lookup queue.
+        assertThat(queue.getReEnqueuedLookupQueue()).hasSize(1);
+        assertThat(queue.hasUnDrained()).isTrue();
+
+        assertThat(queue.drain()).hasSize(5);
+        // drain re-enqueued lookup first.
+        assertThat(queue.getReEnqueuedLookupQueue().isEmpty()).isTrue();
+        assertThat(queue.getLookupQueue()).hasSize(1);
+        assertThat(queue.hasUnDrained()).isTrue();
+
+        assertThat(queue.drain()).hasSize(1);
         assertThat(queue.hasUnDrained()).isFalse();
     }
 

